src/chef-server-ctl/plugins/maintenance.rb

```diff
@@ -0,0 +1,76 @@
+
+require 'redis'
+require "resolv"
+
+def redis
+  @redis ||= begin
+               vip = running_config["private_chef"]["redis_lb"]["vip"]
+               port = running_config["private_chef"]["redis_lb"]["port"]
+               password = credentials.get('redis_lb', 'password')
+               Redis.new(port: port, ip: vip, password: password)
+             end
+end
+
+def enable_maintenance
+  puts "- Enabling maintenance mode"
+  redis.hset("dl_default", "503_mode", true)
+end
+
+def disable_maintenance
+  puts "- Disabling maintenance mode"
+  redis.hdel("dl_default", "503_mode")
+end
+
+def add_ip_to_whitelist(ip)
+  validate_ip(ip)
+  puts "- Adding #{ip} to whitelist."
+  redis.sadd("maint_whitelisted_ip", ip)
+end
+
+def remove_ip_from_whitelist(ip)
+  validate_ip(ip)
+  puts "- Removing #{ip} from whitelist."
+  redis.srem("maint_whitelisted_ip", ip)
+end
+
+def validate_ip(ip)
+  isIPv4 = ip =~ Resolv::IPv4::Regex ? true : false
+  isIPv6 = ip =~ Resolv::IPv6::Regex ? true : false
+  unless isIPv4 or isIPv6
+    $stderr.puts "The IP address is invalid, please enter valid IP address"
+    exit 1
+  end
+end
```

Comment:
This is something you could write a unit test for.

Reply:
Im not sure if have for the functions. But there are test case for the command under spec. Have add the test case there.